NewsApp/Controllers/WeatherViewController/WeatherViewController.swift

```diff
@@ -0,0 +1,438 @@
+        } else if status == .authorizedWhenInUse || status == .authorizedAlways {
+            // Location access was granted, start updating location
+            locationManager.startUpdatingLocation()
+        }
+    }
+    
+    func updateSearchResults(for searchController: UISearchController) {
+        guard let query = searchController.searchBar.text, !query.isEmpty else { return }
+        
+        searchWorkItem?.cancel()
+        
+        let workItem = DispatchWorkItem { [weak self] in
+            self?.fetchWeatherData(for: query)
+        }
+        searchWorkItem = workItem
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5, execute: workItem) // 0.5-second debounce
+    }
+    
+    private func fetchWeatherData(for lat: Double, lon: Double) {
+        print("Fetching weather data for coordinates: lat = \(lat), lon = \(lon)")
+        NewsService.shared.fetchWeatherData(latitude: lat, longitude: lon) { [weak self] result in
+            switch result {
+            case .success(let weatherModel):
+                DispatchQueue.main.async {
+                    if weatherModel.cod == "404" {
+                        self?.showError("The city may not have weather data available.")
+                    } else {
+                        self?.errorLabel.isHidden = true
+                        self?.updateUI(with: weatherModel)
+                    }
+                }
+            case .failure(let error):
+                DispatchQueue.main.async {
+                    if let urlError = error as? URLError {
+                        self?.showError("Network error. Please check your internet connection.")
+                    } else {
+                        self?.showError("Error fetching weather: \(error.localizedDescription). Try searching for another location.")
+                    }
+                }
+            }
+        }
+    }
+    
+    private func fetchWeatherData(for city: String) {
+        NewsService.shared.fetchWeatherData(city: city) { [weak self] result in
+            switch result {
+            case .success(let weatherModel):
+                DispatchQueue.main.async {
+                    if weatherModel.cod == "404" {
+                        self?.showError("The city may not have weather data available.")
+                    } else {
+                        self?.errorLabel.isHidden = true
+                        self?.updateUI(with: weatherModel)
+                    }
+                }
+            case .failure(let error):
+                DispatchQueue.main.async {
+                    if let urlError = error as? URLError {
+                        self?.showError("Network error. Please check your internet connection.")
+                    } else {
+                        self?.showError("Error fetching weather: \(error.localizedDescription). Try searching for another location.")
+                    }
+                }
+            }
+        }
+    }
```

Comment:
You can move all these API calling functions to a ViewModel. Check out the `NewsViewModel`.